waiter/src/waiter/service_description.clj

```diff
@@ -439,21 +456,38 @@
   (compute-profile-defaults
     user-metadata-keys token-defaults profile->defaults profile))
 
+(defn apply-param-to-param-defaults
+  "Returns a service description where any parameter that is a key in the given parameter
+   default mapping and has no value in the service description will be given a value from
+   another parameter in the service description (the parameter name that the original
+   parameter name maps to in the parameter default mapping)."
+  [service-description-defaults param-default-mapping]
+  (let [valid-mapping-parms (map key (filter (fn [[_ mapped-param]] (contains? service-description-defaults mapped-param)) param-default-mapping))
+        new-params (into (sorted-set) (concat (keys service-description-defaults) valid-mapping-parms))]
+    (pc/map-from-keys (fn [param-name] (or (get service-description-defaults param-name)
+                                           (get service-description-defaults (get param-default-mapping param-name))))
+                      new-params)))
+
 (defn merge-defaults
   "Merges the defaults into the existing service description."
-  [{:strs [profile] :as service-description-without-defaults} service-description-defaults profile->defaults metric-group-mappings]
-  (-> service-description-defaults
-    (compute-service-defaults profile->defaults profile)
-    (adjust-default-min-instances service-description-without-defaults)
-    (merge-parameters service-description-without-defaults)
-    (metric-group-filter metric-group-mappings)))
+  ([service-description-without-defaults service-description-defaults profile->defaults metric-group-mappings]
+   (merge-defaults service-description-without-defaults service-description-defaults profile->defaults metric-group-mappings {}))
+  ([{:strs [profile] :as service-description-without-defaults} service-description-defaults profile->defaults metric-group-mappings param-default-mapping]
+   (-> service-description-defaults
+       (compute-service-defaults profile->defaults profile)
+       (adjust-default-min-instances service-description-without-defaults)
+       (merge-parameters service-description-without-defaults)
+       (metric-group-filter metric-group-mappings)
+       (apply-param-to-param-defaults param-default-mapping))))
 
 (defn default-and-override
   "Adds defaults and overrides to the provided service-description"
-  [service-description service-description-defaults profile->defaults metric-group-mappings kv-store service-id]
-  (-> service-description
-      (merge-defaults service-description-defaults profile->defaults metric-group-mappings)
-      (merge-overrides (:overrides (service-id->overrides kv-store service-id)))))
+  ([service-description service-description-defaults profile->defaults metric-group-mappings kv-store service-id]
+   (default-and-override service-description service-description-defaults profile->defaults metric-group-mappings kv-store service-id {}))
```

Comment:
Why is the default to use an empty map?

Reply:
This multi-arity was added for unit testing purposes. In the actual code, the default/empty map path is not followed.

Reply:
Can we avoid the multiple-arity altogether if production code does not use it? We can modify the tests as needed.